ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java

```diff
@@ -2572,18 +2569,24 @@ public static ContentSummary getInputSummary(final Context ctx, MapWork work, Pa
 
       int numExecutors = getMaxExecutorsForInputListing(ctx.getConf(), pathNeedProcess.size());
       if (numExecutors > 1) {
```

Comment:
The default value of `numExecutors` is 1 unless `pathNeedProcess.size()` > 1 and `hive.exec.input.listing.max.threads` sets to the value bigger than 1. We do not need the global lock when `numExecutors = 1`,  and can be converted to a finer-grained lock. In our cluster, we found that many handler threads were blocked by the global lock when one of federation namespaces failover.

Reply:
I am not sure that the problem is with our threads, and not with threads calling this method

Reply:
The problem is that the static global block `INPUT_SUMMARY_LOCK` will block other threads that calling this method.  The `INPUT_SUMMARY_LOCK`  is used for avoiding number of threads out of control, but this method does not use thread pool to get the summaries by default, so the lock can be converted to a finer-grained lock, which only takes care of the creation and execution of the thread pool.

Reply:
Thanks for the explanation!
This seems reasonable!